pkg/controller/nodelifecycle/scheduler/taint_manager.go

```diff
@@ -338,16 +340,24 @@ func (tc *NoExecuteTaintManager) cancelWorkWithEvent(nsName types.NamespacedName
 }
 
 func (tc *NoExecuteTaintManager) processPodOnNode(
-	podNamespacedName types.NamespacedName,
+	pod *v1.Pod,
 	nodeName string,
-	tolerations []v1.Toleration,
 	taints []v1.Taint,
 	now time.Time,
 ) {
+	podNamespacedName := types.NamespacedName{Namespace: pod.Namespace, Name: pod.Name}
 	if len(taints) == 0 {
 		tc.cancelWorkWithEvent(podNamespacedName)
 	}
-	allTolerated, usedTolerations := v1helper.GetMatchingTolerations(taints, tolerations)
+	// Pod will be modified, so making copy is required.
+	podCopy := pod.DeepCopy()
+	// Set reason and message in the pod object.
+	if _, err := nodeutil.SetPodTerminationReason(tc.client, podCopy, nodeName); err != nil {
```

Comment:
I don't see how `NodeLost` status (defined as `NodeUnreachablePodReason is the reason on a pod when its state cannot be confirmed as kubelet is unresponsive on the node it is (was) running.`) is fully related to a pod being evicted due to NoExecute taints no longer tolerated or not tolerated given a timeout.

I can imagine the case where the node controllers realizes a node is no longer posting its status. Thus, it decides to evict all the pods on the node by setting a taint with NoExecute as is described in [1]: `In case a node is to be evicted, the node controller or the kubelet adds relevant taints with NoExecute effect.`. So at least pods have a chance to report something "bad" happened. Though, in case a pod extend its list of tolerations (which is a sufficient condition for a pod getting updated and `helper.Semantic.DeepEqual(oldTolerations, newTolerations)` [2] no longer being true), with this PR the pod's status reason gets set to `NodeLost` status even though the node is not lost.

Maybe I just don't see the full picture how this code changes fixes https://github.com/kubernetes/kubernetes/issues/89212. @chenkaiyue can you more elaborate on how this fixes the issue?

[1] https://kubernetes.io/docs/concepts/scheduling-eviction/taint-and-toleration/#taint-based-evictions
[2] https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/nodelifecycle/scheduler/taint_manager.go#L297

Reply:
https://github.com/kubernetes/kubernetes/blob/d66bbd87283fd5219f84e89ccafb6cffe3396171/pkg/controller/nodelifecycle/node_lifecycle_controller.go#L1320 
https://github.com/kubernetes/kubernetes/blob/d66bbd87283fd5219f84e89ccafb6cffe3396171/pkg/controller/nodelifecycle/node_lifecycle_controller.go#L909 https://github.com/kubernetes/kubernetes/blob/d66bbd87283fd5219f84e89ccafb6cffe3396171/pkg/controller/nodelifecycle/node_lifecycle_controller.go#L924
https://github.com/kubernetes/kubernetes/blob/d66bbd87283fd5219f84e89ccafb6cffe3396171/pkg/controller/nodelifecycle/node_lifecycle_controller.go#L1463

When taint-based eviction is not used, when node ready condition is ConditionFalse or Unknown, if a node has NoExecute taint, the pod on it will be marked by `SetPodTerminationReason`. So the logic here is the same. Maybe I should add codes later about judging whether node status is ConditionFalse or Unknown as mentioned above and keep pace with above codes.

https://github.com/kubernetes/kubernetes/blob/d66bbd87283fd5219f84e89ccafb6cffe3396171/pkg/controller/nodelifecycle/node_lifecycle_controller.go#L866
But from here, we know that only when node ready condition is ConditionFalse or Unknown, nodes will be added TaintEffectNoExecute(TaintNodeUnreachable or TaintNodeNotReady).

> in case a pod extend its list of tolerations (which is a sufficient condition for a pod getting updated and helper.Semantic.DeepEqual(oldTolerations, newTolerations) [2] no longer being true), with this PR the pod's status reason gets set to NodeLost status even though the node is not lost.

This will not happen, because in NoExecuteTaintManager, `nodeupdate` will run first https://github.com/kubernetes/kubernetes/blob/d66bbd87283fd5219f84e89ccafb6cffe3396171/pkg/controller/nodelifecycle/scheduler/taint_manager.go#L256 ,then is `handlePodUpdate` https://github.com/kubernetes/kubernetes/blob/d66bbd87283fd5219f84e89ccafb6cffe3396171/pkg/controller/nodelifecycle/scheduler/taint_manager.go#L271 , and in `handlePodUpdate` we can see only if node is in `tc.taintedNodesnoExecute` https://github.com/kubernetes/kubernetes/blob/d66bbd87283fd5219f84e89ccafb6cffe3396171/pkg/controller/nodelifecycle/scheduler/taint_manager.go#L412 , pod will have chance to be marked by `NodeUnreachablePodReason`. If pods only extend its list of tolerations and node doesn't have taints, NoExecuteTaintManager will not mark pod, after all NoExecuteTaintManager works for situation about NoExecuteTaint.

Reply:
> This will not happen, because in NoExecuteTaintManager, nodeupdate will run first

`handleNodeUpdate` gets called only when a node gets updated. I am concerned about the case when only a pod gets updated, node's taints does not change. E.g. A node has taint Ta1 which is tolerated by pod toleration To1 without a timeout. At some point, To1 gets updated, its timeout gets set to 5 minutes. At that point, your change mark the pod's state as `NodeLost` even though node taint Ta1 has not changed. Yet, the node is not lost.